src/ring_theory/noetherian.lean

```diff
@@ -199,6 +200,45 @@ begin
     { exact zero_smul _ }, { exact λ _ _ _, add_smul _ _ _ } }
 end
 
+lemma singleton_span_is_compact_element (x : M) :
+  complete_lattice.is_compact_element (span R {x} : submodule R M) :=
+begin
+  rw complete_lattice.is_compact_element_iff_le_of_directed_Sup_le,
+  intros d hemp hdir hsup,
+  have : x ∈ Sup d, from (le_def.mp hsup) (mem_span_singleton_self x),
+  obtain ⟨y, ⟨hyd, hxy⟩⟩ := (mem_Sup_of_directed hemp hdir).mp this,
+  exact ⟨y, ⟨hyd, by simpa only [span_le, singleton_subset_iff]⟩⟩,
+end
+
+/-- Finitely generated submodules are precisely compact elements in the submodule lattice -/
+theorem fg_iff_compact (s : submodule R M) : s.fg ↔ complete_lattice.is_compact_element s :=
+begin
+  classical,
+  -- Introduce shorthand for span of an element
+  let sp : M → submodule R M := λ a, span R {a},
+  split,
+  { rintro ⟨t, rfl⟩,
+    rw span_eq_sup_of_singleton_spans,
+    change complete_lattice.is_compact_element (⨆ (x : M) (H : x ∈ t), span R {x}),
+    rw ←(finset.sup_eq_supr t sp),
```

Comment:
I'm still unhappy with these three lines, and their identical counterpart in the converse.
In both cases, I want to go between
`(⨆ (x : M) (H : x ∈ ↑t), span R {x})` and `t.sup sp`. I've tried all sorts of `simp` or `simp_rw` strategies, but `simp_rw` seems really sensitive to getting stuck, and `squeeze_simp` gives suggestions that don't work. At the end of the day I'd like to avoid the manual `change` if at all possible.

Reply:
I haven't checked in Lean, but is there a lemma you could add to rewrite with?

Reply:
It's a bit awkward because, in both places, all the "content" is essentially taken care of by `sup_eq_supr`. In both cases, the chain of rewrites needs to be broken by manually changing 
```
⨆ (x : M) (H : x ∈ ↑t), span R {x}
```
to
```
⨆ (x : M) (H : x ∈ t), span R {x}
```
Where the only difference is the `coe`. For removing the `coe`, I can't get `simp` or `simp_rw` to do it, and I asked about this in the chat. For putting the coe back in, of course it wouldn't be done by simp, so I don't know here.

I suppose I could write a lemma for this _specific_ change, but it would just be `by refl`, so it feels a little weird.